In [1]:
import torch
from torchsummary import summary

#### Homework (see Moodle for additional info)

1. build the MLP using PT built-ins as above
2. Instantiate and summarise
3. Provide calculation for the exact number of parameters of the MLP (also in the case of bias terms)
4. Calculate the L1 and L2 norm of parameters for the params of each layer

In [2]:
class MLP_New(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.layer1 = torch.nn.Linear(in_features=5, out_features=11,bias=False)
        self.layer2 = torch.nn.Linear(in_features=11, out_features=16,bias=False)
        self.layer3 = torch.nn.Linear(in_features=16, out_features=13,bias=False)
        self.layer4 = torch.nn.Linear(in_features=13, out_features=8,bias=False)
        self.layer5 = torch.nn.Linear(in_features=8, out_features=4,bias=False)

    def forward(self, X):
        out = self.layer1(X)
        out = torch.nn.functional.relu(out)
        out = self.layer2(out)
        out = torch.nn.functional.relu(out)
        out = self.layer3(out)
        out = torch.nn.functional.relu(out)
        out = self.layer4(out)
        out = torch.nn.functional.relu(out)
        out = self.layer5(out)
        out = torch.nn.functional.softmax(out)
        return out

    def getListOfLayers(self):
        return [self.layer1,self.layer2,self.layer3,self.layer4,self.layer5]

    def printNormsOfLayers(self,norm_type):        
        #list_of_layers = self.getListOfLayers()
        print("L{norm_type} norm of layer1 is:".format(norm_type,torch.norm(self.layer1)))

In [3]:
mlp_model = MLP_New()
mlp_model

MLP_New(
  (layer1): Linear(in_features=5, out_features=11, bias=False)
  (layer2): Linear(in_features=11, out_features=16, bias=False)
  (layer3): Linear(in_features=16, out_features=13, bias=False)
  (layer4): Linear(in_features=13, out_features=8, bias=False)
  (layer5): Linear(in_features=8, out_features=4, bias=False)
)

In [4]:
summary(mlp_model)

Layer (type:depth-idx)                   Param #
├─Linear: 1-1                            55
├─Linear: 1-2                            176
├─Linear: 1-3                            208
├─Linear: 1-4                            104
├─Linear: 1-5                            32
Total params: 575
Trainable params: 575
Non-trainable params: 0


Layer (type:depth-idx)                   Param #
├─Linear: 1-1                            55
├─Linear: 1-2                            176
├─Linear: 1-3                            208
├─Linear: 1-4                            104
├─Linear: 1-5                            32
Total params: 575
Trainable params: 575
Non-trainable params: 0

In [5]:
# Below part is calculated by manually, hope it was the purpose of question
print("Number of parameters without bias: ",5*11 + 11*16 + 16*13 + 13*8 + 8*4)
print("Number of parameters with bias: ",5*11 + 11 + 11*16 + 16 + 16*13 + 13 + 13*8 + 8+ 8*4 + 4)

Number of parameters without bias:  575
Number of parameters with bias:  627


* Number of parameters without bias: $5 \times 11 + 11 \times 16 + 16 \times 13 + 13 \times 8 + 8 \times 4 = 575$

* Number of parameters with bias: $5 \times 11 + 11 + 11 \times 16 + 16 + 16 \times 13 + 13 + 13 \times 8 + 8 + 8 \times 4 + 4 = 627$

In [6]:
model_layers_dict = mlp_model.state_dict()
for layer_name, layer_weigths in model_layers_dict.items():
    print(layer_name, "L1 norm of layer is: ", torch.linalg.norm(layer_weigths,1), "L2 norm of layer is: ", torch.linalg.norm(layer_weigths,2))

layer1.weight L1 norm of layer is:  tensor(3.1176) L2 norm of layer is:  tensor(1.4463)
layer2.weight L1 norm of layer is:  tensor(3.0540) L2 norm of layer is:  tensor(1.1616)
layer3.weight L1 norm of layer is:  tensor(2.3010) L2 norm of layer is:  tensor(1.0164)
layer4.weight L1 norm of layer is:  tensor(1.3139) L2 norm of layer is:  tensor(0.9169)
layer5.weight L1 norm of layer is:  tensor(0.7274) L2 norm of layer is:  tensor(0.6791)
